### PR TITLE
Resolution fix

### LINKED ISSUE #45: Settings Menu: Add resolution options
Split from #42

### PR DESCRIPTION
**Pull request by yaelatletl**
Works upon #61 and closes  #45

### CODE REVIEW

File: addons/HuePicker/HuePickerUtils.gd [hunk 1]
```diff
@@ -1,9 +1,10 @@
-func HSVtoRGB (hue, saturation, luminance):
-	var r = float(luminance)
-	var g = float(luminance)
-	var b = float(luminance)
+# TODO: Use class_name so this can be called without instancing
+func HSVtoRGB (hue : float, saturation : float, luminance: float) -> Color:
```
Comment (TheDuriel) [inline addons/HuePicker/HuePickerUtils.gd]:
```
vec3 hsv2rgb(vec3 c) {
	vec4 K = vec4(1.0, 2.0 / 3.0, 1.0 / 3.0, 3.0);
	vec3 p = abs(fract(c.xxx + K.xyz) * 6.0 - K.www);
	return c.z * mix(K.xxx, clamp(p - K.xxx, 0.0, 1.0), c.y);
}

vec3 rgb2hsv(vec3 c)
{
	vec4 K = vec4(0.0, -1.0 / 3.0, 2.0 / 3.0, -1.0);
	vec4 p = mix(vec4(c.bg, K.wz), vec4(c.gb, K.xy), step(c.b, c.g));
	vec4 q = mix(vec4(p.xyw, c.r), vec4(c.r, p.yzx), step(p.x, c.r));

	float d = q.x - min(q.w, q.y);
	float e = 1.0e-10;
	return vec3(abs(q.z + (q.w - q.y) / (6.0 * d + e)), d / (q.x + e), q.x);
}
```
Add these as a comment. Or convert them yourself. This is how you do the conversion in shader code.

File: addons/HuePicker/PickerPanel.gd [hunk 2]
```diff
@@ -1,89 +1,78 @@
...
 
-export(bool) var flat setget flat_changed
+export(Color) var color : Color setget set_color
+export(bool) var flat : bool  setget set_flat
+
+onready var sliders = $ClassicControls/Hider/Viewport/ColorPicker
+onready var HuePicker = $HuePicker
 
-var isReady = false
+var isReady : bool = false
```
Comment (TheDuriel) [inline addons/HuePicker/PickerPanel.gd]: is_ready

also this seems ugly. leave a FIXME comment please.

File: addons/HuePicker/clip.gd [hunk 3]
```diff
@@ -72,3 +37,13 @@ func update_shaders():
 	$'../B_Prev'.material.set_shader_param("color2", Color(color.r,color.g,1,1))
 	$'../A_Prev'.material.set_shader_param("color1", Color(color.r,color.g,color.b,0))
 	$'../A_Prev'.material.set_shader_param("color2", Color(color.r,color.g,color.b,1))
+
+func _on_ColorPicker_focus_entered() -> void:
+	Focused = true 
+func _on_ColorPicker_focus_exited() -> void:
+	Focused = false
+
+##  UPDATE THE HUE CIRCLE BASE COLOR ETC.
```
Comment (TheDuriel) [inline addons/HuePicker/clip.gd]: make this a TODO or FIXME

File: addons/HuePicker/clip.gd [hunk 4]
```diff
@@ -1,69 +1,34 @@
...
-##		$'../../..'._on_HuePicker_color_changed(color)
-#		$'../../..'.emit_signal('color_changed', color)
-#
-#	$'../../..'.color.a = color.a
-
-	#update the classic controls.
-	update_shaders()
-	
-func update_shaders():
+func update_shaders() -> void:
```
Comment (TheDuriel) [inline addons/HuePicker/clip.gd]: TODO: Fix these paths.

File: addons/HuePicker/clip2.gd [hunk 5]
```diff
@@ -2,54 +2,58 @@
...
 signal color_changed(color)
 
-var isReady = false
+export(Color) var color : Color setget color_changed
+
+onready var picker = $Hider/Viewport/ColorPicker
+onready var Hider = $Hider
+onready var viewport : Viewport = $Hider/Viewport
+
+var isReady : bool = false
```
Comment (TheDuriel) [inline addons/HuePicker/clip2.gd]: Same as with the other one.

File: addons/HuePicker/indicator_rgba.gd [hunk 6]
```diff
@@ -1,8 +1,11 @@
 extends PanelContainer
 
-func _gui_input(event):
+onready var pop = $Pop
+onready var Color_picker = $Pop/ColorPicker
```
Comment (TheDuriel) [inline addons/HuePicker/indicator_rgba.gd]: color_picker

File: assets/Player/avatar_v2/scripts/Player.gd [hunk 7]
```diff
@@ -128,10 +128,10 @@ func SetPuppetColors(var colors):
 	shoes_mat.albedo_color = colors.shoes
 
 func SetPuppetGender(var gender):
-	$KinematicBody/Model/FemaleRig/Skeleton/AvatarFemale.visible = (gender == options.genders.female)
```
Comment (TheDuriel) [inline assets/Player/avatar_v2/scripts/Player.gd]: Paths. ;P

File: assets/UI/Menu/Instructions/Resizing.gd [hunk 8]
```diff
@@ -1,8 +1,9 @@
 extends ScrollContainer
```
Comment (TheDuriel) [inline assets/UI/Menu/Instructions/Resizing.gd]: I am rather confused by this. and I am nearly certain that this script is completely obsolete, and can instead be replaced with an anchors and margin preset. I'll make an issue.

File: assets/UI/Menu/lobby.gd [hunk 9]
```diff
@@ -22,6 +21,11 @@ func _ready():
 # 	gamestate.connect("game_ended", self, "_on_game_ended")
 # 	gamestate.connect("game_error", self, "_on_game_error")
 
+func set_name(name : String = "") -> void:
+	if name == "":
+		name = namelist.get_name()
+	$connect/name.text = name
```
Comment (TheDuriel) [inline assets/UI/Menu/lobby.gd]: Paths ;P

File: assets/UI/Menu/lobby.gd [hunk 10]
```diff
@@ -66,8 +79,34 @@ func sg_network_error(msg):
 		STATE.CLIENT_CONNECT:
 			$connect/error_label.text = msg
 
-func sg_server_connected():
+func sg_server_connected() -> void:
 	sg_server_up()
+	
+func _bindsg(_signal, _sub = null):
```
Comment (TheDuriel) [inline assets/UI/Menu/lobby.gd]: Add a TODO comment: Rename everything.

File: assets/UI/Options/Options.gd [hunk 11]
```diff
@@ -68,91 +72,73 @@ func _get_player():
...
-func init_playerspeed_control(button):
-	var player = _get_player()
+func init_playerspeed_control(button : Control) -> void:
+	var player = get_player()
 	if player:
 		button.enabled = true
 		button.value = player.get("SPEED_SCALE")
 		button.connect("changed", self, "set_player_speed")
-		printd("found player %s at %s, enable speed changes" % [player, player.get_path()])
+		print("found player", player, "at", player.get_path(), ", enable speed changes")
```
Comment (TheDuriel) [inline assets/UI/Options/Options.gd]: prints() for adding spaces between values.

File: assets/UI/Options/Settings.gd [hunk 12]
```diff
@@ -35,23 +50,24 @@ func _ready() -> void:
 			current_resolution = resolution_index
 	
 	if current_resolution == -1:
-		Resolution.get_popup().add_item(str(res_width, " x ", res_height))
-		resolutions.append(Vector2(res_width, res_height))
+		Resolution.get_popup().add_item(str(OS.window_size.x, " x ", OS.window_size.y))
+		resolutions.append(Vector2(OS.window_size.x, OS.window_size.y))
```
Comment (TheDuriel) [inline assets/UI/Options/Settings.gd]: OS.window_size is already a vector2

File: assets/UI/Options/Settings.gd [hunk 13]
```diff
@@ -35,23 +50,24 @@ func _ready() -> void:
...
 	
+	ScreenMode.text = mode
 	Resolution.text = Resolution.get_popup().get_item_text(current_resolution)
 
 func _on_Detail_change(id : int) -> void:
 	Quality.text = Quality.get_popup().get_item_text(id)
 
 func _on_Resolution_change(id : int) -> void:
+	if current_mode == 0:
+		OS.window_size = Vector2(resolutions[current_resolution].x, resolutions[current_resolution].y)
```
Comment (TheDuriel) [inline assets/UI/Options/Settings.gd]: already stored as a Vector2, no need to reconstruct it.

File: core/autoload/autoload-options.gd [hunk 14]
```diff
@@ -53,8 +57,8 @@ var shoes_color : Color = Color(0,0,0,1)
 var savefile_json
 
 #FIXME Writing and reading should be done with a ConfigFile instead of writing a dictionary using File.
-var OptionsFile : String = "user://gameoptions.save"
-var User_file : String = "user://settings.save"
+var OptionsFile : String = "res://gameoptions.save"
```
Comment (TheDuriel) [inline core/autoload/autoload-options.gd]: options_file_path

user_file_path

File: core/autoload/autoload-options.gd [hunk 15]
```diff
@@ -190,6 +194,7 @@ func _ready() -> void:
 	self.load()
 	set_defaults()
 	LoadUserSettings()
+	LoadGraphicSettings()
```
Comment (TheDuriel) [inline core/autoload/autoload-options.gd]: snake case please

File: core/autoload/autoload-options.gd [hunk 16]
```diff
@@ -343,3 +348,23 @@ func SaveUserSettings():
 	savefile.store_line(to_json(save_dict))
 	savefile.close()
 	emit_signal("user_settings_changed")
+	
+func LoadGraphicSettings():
```
Comment (TheDuriel) [inline core/autoload/autoload-options.gd]: This is all duplicate code from Settings.gd it seems. I'll make an issue.

Comment (TheDuriel) [reply core/autoload/autoload-options.gd]: #90

File: assets/UI/Options/Settings.gd [hunk 17]
```diff
@@ -60,11 +76,26 @@ func _on_UpdateCheck_pressed() -> void:
 	else:
 		options.set("updater/client", false, "check_at_startup")
 
+func _on_ScreenMode_change(id : int) -> void:
```
Comment (TheDuriel) [inline assets/UI/Options/Settings.gd]: #90 and the rest of this file. NOT part of this pull request.

File: assets/UI/Menu/Instructions/Resizing.gd (hunk 8 above)
Comment (TheDuriel) [reply assets/UI/Menu/Instructions/Resizing.gd]: #91

**Comment by yaelatletl** [comment]
I'm not sure what happened here, but there are files that are not part of this PR. Only settings.gd, project.godot and autoload-options are part of the PR

File: core/autoload/autoload-options.gd (hunk 16 above)
Comment (yaelatletl) [reply core/autoload/autoload-options.gd]: Needs to be here for it to update at run-time, the fix would belong to settings.gd

**Comment by TheDuriel** [comment]
You did not rebase and branch correctly. Same problem with the other PR.

image

Blue is moonwards1/master, which is what you should be branching from. Your own master should follow it identically, and never be committed to.
